spartan_parallel/src/scalar/mod.rs

```diff
@@ -1,43 +1,325 @@
-mod ristretto255;
+mod fp;
+mod fp2;
 
-pub type Scalar = ristretto255::Scalar;
-pub type ScalarBytes = curve25519_dalek::scalar::Scalar;
+use ceno_goldilocks::ExtensionField;
+use ff::Field;
+pub use fp::Scalar;
+pub use fp2::ScalarExt2;
+use merlin::Transcript;
+use rand::{CryptoRng, RngCore};
+use serde::Serialize;
+use std::fmt;
+use std::{
+  cmp::Eq,
+  hash::Hash,
+  iter::{Product, Sum},
+  ops::{Add, Mul, Neg, Sub},
+};
+use subtle::{Choice, ConditionallySelectable, ConstantTimeEq, CtOption};
+use zeroize::Zeroize;
 
-pub trait ScalarFromPrimitives {
-  fn to_scalar(self) -> Scalar;
-}
+/// Trait describing the field element
+/// Wraps around Goldilocks field towers from ceno-goldilocks
+/// See: https://github.com/scroll-tech/ceno-Goldilocks
+pub trait SpartanExtensionField:
+  Sized
+  + ConstantTimeEq
+  + Eq
+  + PartialEq
+  + From<u64>
+  + From<usize>
+  + ConditionallySelectable
+  + Zeroize
+  + Neg
+  + Default
+  + Add<Output = Self>
+  + Sub<Output = Self>
+  + Mul<Output = Self>
+  + Sum
+  + Product
+  + Clone
+  + Serialize
+  + Hash
+  + From<Self::InnerType>
+  + fmt::Debug
+{
+  /// Inner Goldilocks extension field
+  type InnerType: ExtensionField + Field;
```

Comment:
It's pretty cumbersome to access its base field via `<S::InnerType as ExtensionField>::BaseField`, right?

Reply:
We need to use BaseField in a lot of places for performance. For example, the type of each cell's value in the R1CS matrices must belong to the base field.